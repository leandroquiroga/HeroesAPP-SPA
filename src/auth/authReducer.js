import { types } from "../types/type";



export const authReducer = (state = {}, action) => {
  const { type, payload } = action;
  
  switch (type) {
    case types.login:
      return {
        ...payload,
        logged: true,
      };
    case types.logout:
      return {
        logged: false,
      };
    
    default:
      return state;
  }
}