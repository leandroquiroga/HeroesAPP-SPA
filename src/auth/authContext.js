import { createContext, useEffect, useReducer, useState } from "react";
import { signInWithEmailAndPassword, signOut ,createUserWithEmailAndPassword } from 'firebase/auth';
import { authReducer } from "./authReducer";
import { auth } from './../firebase';
import { heroReducer } from "../helpers";

// Retorna el usuario si lo encuentra en localStorage si no retorna el logged en false
const init = () => JSON.parse(localStorage.getItem('user')) || {logged: false}

// Retorna los hero si lo encuntra en localStorage si no retorna un array 
const initHero = () => JSON.parse(localStorage.getItem('hero')) || [];

export const AuthContext = createContext();

export const AuthProvaider = (props) => {
  const [user, dispatch] = useReducer(authReducer, {}, init);
  const [currentUser, setCurrentUser] = useState({});
  const [hero, dispatchHero] = useReducer(heroReducer, [], initHero);

  // Cuando inicia la aplicacion va a estar observando si hay un cambio
  // de usuario.
  useEffect(() => {
    auth.onAuthStateChanged((user) => setCurrentUser(user))
  }, []);

  // Esta pendiente al user para ir almancendolo en el localStorage
  useEffect(() => {
    localStorage.setItem('user', JSON.stringify(user))
    return () => {
      if (!user) return;
    }
  }, [user]);

  // Esta pendiente al hero, cuando detecta un cambio lo almacena en el localStorage
  useEffect(() => {
    localStorage.setItem('hero', JSON.stringify(hero));
    return () => {
      if (!hero) return;
    }
  }, [hero])
  


  const singUp = (email, password) => createUserWithEmailAndPassword(auth, email, password);

  const logIn = (email, password) => signInWithEmailAndPassword(auth, email, password);
  
  const logOut = (auth) => signOut(auth);

  const value = {
    user,
    dispatch,
    singUp,
    logIn,
    logOut,
    currentUser,
    dispatchHero,
    hero,
  }
  return (
    <AuthContext.Provider value={value}>
      {props.children}
    </AuthContext.Provider>
  )
}