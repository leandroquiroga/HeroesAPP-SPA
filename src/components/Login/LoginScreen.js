import React, { useContext, useState } from 'react';
  
import { Link, useNavigate } from 'react-router-dom';
import { AuthContext } from './../../auth/authContext';
import { useForm } from '../../hooks/useForm';
import { loginUser } from '../../helpers';

export const LoginScreen = () => {
  const [state, handleChangeInput] = useForm({
    emailUser: '',
    password: ''
  });
  const { emailUser, password } = state;
  const { dispatch, currentUser, logIn } = useContext(AuthContext);
  const [error, setError] = useState('')
  const navigate = useNavigate();

  //inida a que pagina se debe regresar
  const handleLogin = (e) => {
    e.preventDefault();
    
    // Valida que los campos no esten vacios
    if ([emailUser, password].includes('')) {
      setError('Completa todos los campos');
      setTimeout(() => setError(''), 2500); 
      return;
    };
    // Autenticacion y busca en el localStorage el ultimo path. 
    loginUser(logIn, dispatch, navigate, setError, state, currentUser);
  }
  return (
    <div className='container-background'>
      <div className='d-flex justify-content-center align-items-center p-2'>
        <div
          className='shadow-lg p-3 mb-5 container-blur rounded text-white'
          style={{ width: 370 + 'px'}}
        >
          <h1 className='text-center fs-4 text-white'>Iniciar sesión</h1>
          <hr />
          <form
            className='form-control border-0 bg-transparent text-white'
            onSubmit={handleLogin}
          >
            <div className='mb-3'>
              <label
                htmlFor='emailUser'
                className='form-label'
              >
                Email :
              </label>
              <input 
                autoFocus
                className='form-control border-0 border-bottom bg-transparent text-white'
                type='email'
                id='emailUser'
                placeholder='Ingrese e-mail'
                name='emailUser'
                value={emailUser}
                onChange={handleChangeInput}
              />
            </div>

            <div className='mb-3'>
              <label
                htmlFor='passUser'
                className='form-label'
              >
                Contraseña: 
              </label>
              <input 
                className='form-control border-0 border-bottom bg-transparent text-white'
                type='password'
                id='passUser'
                placeholder='Ingrese contraseña'
                name='password'
                value={password}
                onChange={handleChangeInput}
              />
            </div>
            {error &&
                <div className='alert alert-danger text-center fw-bold my-2 p-1'>
                  {error}
                </div>
            }
            <button 
              className='btn btn-dark w-100 mt-4'
              type='submit'
            >
              Login
            </button>
          </form>
          <p className='text-center fs-6 text-white'>
            ¿No tienes cuenta? Haz click en <br />
            <Link to='/register'> Registrarme </Link>
          </p>
          <p className='fs-6 mt-3 text-center'>
            Todos los derechos reservados <br />
            &copy;Hero App
          </p>
        </div>
      </div>
    </div>
  );
}