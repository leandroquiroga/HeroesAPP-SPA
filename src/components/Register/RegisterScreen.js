import React, { useContext, useState } from 'react';
import { Link, useNavigate } from 'react-router-dom';
import { AuthContext } from '../../auth/authContext';
import { registerUser } from '../../helpers';
import { useForm } from '../../hooks/useForm';

export const RegisterScreen = () => {
  const [state, handleChangeInput] = useForm({
    emailUser: '',
    password: '',
    passwordConfirm: '',
  });
  const { emailUser, password, passwordConfirm } = state;
  const { dispatch, singUp, currentUser } = useContext(AuthContext);
  const navigate = useNavigate();
  const [error, setError] = useState('')

  const handleRegister = (e) => {
    e.preventDefault();

    // Validacion de campos vacios y contraseñas iguales
    if ([emailUser, password, passwordConfirm].includes('')) {
      setError('Completa todos los campos');
      setTimeout(() => setError(''), 2500);
      return
    } else if (password !== passwordConfirm) {
      setError('Las contraseñas no coinciden');
      setTimeout(() => setError(''), 2500);
      return
    }

    console.log(emailUser);

    // Registra un usuario, si los datos son correctos,
    // lo redirije a la pagina principal de la App
    registerUser(singUp, dispatch, navigate, setError, state, currentUser)
  }

  const imagePath = `/assets/svg/avatar.svg`
  return (
    <div className='container mt-3 '>
      <div className='d-flex justify-content-center align-items-center p-3  '>
        <div
          className='shadow-lg p-3 mb-5 bg-body rounded '
          style={{ width: 370 + 'px'}}
        >
          <h1 className='text-center fs-3'>Registrate</h1>
          <hr />
          <div className='d-flex justify-content-center align-items-center p-1'>
            <img 
              className='img-fluid w-25'
              src={imagePath}
              alt='icon'
            />
          </div>
          <form
            className='form-control  border-0'
            onSubmit={handleRegister}
          >
            <div className='mb-2 '>
              <label
                htmlFor='emailUser'
                className='form-label'
              >
                Email :
              </label>
              <input 
                autoFocus
                className='form-control border-0 border-bottom'
                type='email'
                id='emailUser'
                placeholder='Ingrese e-mail'
                name='emailUser'
                value={emailUser}
                onChange={handleChangeInput}
              />
            </div>

            <div className='mb-2 '>
              <label
                htmlFor='passUser'
                className='form-label'
              >
                Contraseña: 
              </label>
              <input 
                className='form-control border-0 border-bottom'
                type='password'
                id='passUser'
                placeholder='Ingrese contraseña'
                name='password'
                value={password}
                onChange={handleChangeInput}
              />
            </div>
            <div className='mb-2 '>
              <label
                htmlFor='passConfirm'
                className='form-label'
              >
                Repite contraseña: 
              </label>
              <input 
                className='form-control border-0 border-bottom'
                type='password'
                id='passConfirm'
                placeholder='Repite contraseña'
                name='passwordConfirm'
                value={passwordConfirm}
                onChange={handleChangeInput}
              />
            </div>
            {error &&
                <div className='alert alert-danger text-center fw-bold my-2 p-1'>
                  {error}
                </div>
            }
            <button 
              className='btn btn-dark w-100 mt-4'
            >
              Registrate
            </button>
          </form>
          <p className='text-center'>
            ¿Ya eres usuario? Haz click en <br />
            <Link to='/login'> Iniciar Sesión </Link>
          </p>
          <p className='fs-6 mt-3 text-center'>
            Todos los derechos reservados <br />
            &copy;Hero App
          </p>
        </div>
      </div>
    </div>
  );
};
