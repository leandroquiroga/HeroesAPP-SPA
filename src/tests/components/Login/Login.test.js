
import { mount} from 'enzyme';
import { LoginScreen } from './../../../components/Login/LoginScreen';
import { AuthContext } from './../../../auth/authContext';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import { RegisterScreen } from './../../../components/Register/RegisterScreen';
import { act } from '@testing-library/react';
import * as React from 'react';

// const loginUser = jest.fn();

const mockNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate,
}));

describe('Testing the component <LoginScreen/>', () => {

  const contextValue = {
    dispatch: jest.fn(),
    login: jest.fn(),
    hero: [
      {
        'id': 'dc-batman',
        'superhero': 'Batman',
        'publisher': 'DC Comics',
        'alter_ego': 'Bruce Wayne',
        'first_appearance': 'Detective Comics #27',
        'characters': 'Bruce Wayne'
      },
    ],
    user: {
      email: 'leandro@correo.com',
      password: '123456'
    }
  };



  const wapper = mount(
    <AuthContext.Provider value={contextValue}>
      <MemoryRouter initialEntries={['/login']}>
        <Routes>
          <Route path='/login' element={<LoginScreen />} />
        </Routes>
      </MemoryRouter>
    </AuthContext.Provider>
  );
  test('This test must showing correctly', () => {
    

    expect(wapper).toMatchSnapshot();
    expect(wapper.find('h1').text().trim()).toBe('Iniciar sesi√≥n');
    expect(wapper.find('#emailUser').exists()).toBeTruthy();
    expect(wapper.find('#passUser').exists()).toBeTruthy();

  });

  test('This test must navigate to the <Register /> component by clicking', () => {
    
    expect(wapper.find('a').exists()).toBe(true)

    wapper.find('a').prop('onClick')({
      defaultPrevented: () => { }
    });
    
    const pageRegister = mount(
      <AuthContext.Provider value={contextValue}>
        <MemoryRouter initialEntries={['/register']}>
          <Routes>
            <Route path='/register' element={<RegisterScreen />} />
          </Routes>
        </MemoryRouter>
      </AuthContext.Provider>
    );
    
    expect(pageRegister).toMatchSnapshot();
    expect(pageRegister.find('h1').text().trim()).toBe('Registrate')
  });

  test('This test should check if the inputs are not empty', () => {
    const { user } = contextValue;

    wapper.find('#emailUser').simulate('change', {
      target: {
        name: 'email',
        value: user.email
      }
    });
    
    wapper.find('#passUser').simulate('change', {
      target: {
        name: 'password',
        value: user.password
      }
    });
    act(() => { 
      wapper.find('form').prop('onSubmit')({
        preventDefault: () => { }
      });
    })
    

    expect(wapper.find('#emailUser').prop('value')).toBe(user.email)
    expect(wapper.find('#passUser').prop('value')).toBe(user.password);
    
  });

})