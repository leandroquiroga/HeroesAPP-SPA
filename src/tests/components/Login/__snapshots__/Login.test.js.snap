// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the component <LoginScreen/> This test must navigate to the <Register /> component by clicking 1`] = `
<div
  className="container-background"
>
  <div
    className="d-flex justify-content-center align-items-center p-3 text-white"
  >
    <div
      className="shadow-lg p-3 mb-5 container-blur rounded "
      style={
        Object {
          "width": "370px",
        }
      }
    >
      <h1
        className="text-center fs-3"
      >
        Registrate
      </h1>
      <hr />
      <form
        className="form-control border-0 bg-transparent text-white"
        onSubmit={[Function]}
      >
        <div
          className="mb-2 "
        >
          <label
            className="form-label"
            htmlFor="emailUser"
          >
            Email :
          </label>
          <input
            autoFocus={true}
            className="form-control border-0 border-bottom bg-transparent text-white"
            id="emailUser"
            name="emailUser"
            onChange={[Function]}
            placeholder="Ingrese e-mail"
            type="email"
            value=""
          />
        </div>
        <div
          className="mb-2 "
        >
          <label
            className="form-label"
            htmlFor="passUser"
          >
            Contraseña:
          </label>
          <input
            className="form-control border-0 border-bottom bg-transparent text-white"
            id="passUser"
            name="password"
            onChange={[Function]}
            placeholder="Ingrese contraseña"
            type="password"
            value=""
          />
        </div>
        <div
          className="mb-2 "
        >
          <label
            className="form-label"
            htmlFor="passConfirm"
          >
            Repite contraseña:
          </label>
          <input
            className="form-control border-0 border-bottom bg-transparent text-white"
            id="passConfirm"
            name="passwordConfirm"
            onChange={[Function]}
            placeholder="Repite contraseña"
            type="password"
            value=""
          />
        </div>
        <button
          className="btn btn-dark w-100 mt-4"
        >
          Registrate
        </button>
      </form>
      <p
        className="text-center"
      >
        ¿Ya eres usuario? Haz click en 
        <br />
        <a
          href="/login"
          onClick={[Function]}
        >
           Iniciar Sesión 
        </a>
      </p>
      <p
        className="fs-6 mt-3 text-center"
      >
        Todos los derechos reservados 
        <br />
        ©Hero App
      </p>
    </div>
  </div>
</div>
`;

exports[`Testing the component <LoginScreen/> This test must showing correctly 1`] = `
<div
  className="container-background"
>
  <div
    className="d-flex justify-content-center align-items-center p-2"
  >
    <div
      className="shadow-lg p-3 mb-5 container-blur rounded text-white"
      style={
        Object {
          "width": "370px",
        }
      }
    >
      <h1
        className="text-center fs-4 text-white"
      >
        Iniciar sesión
      </h1>
      <hr />
      <form
        className="form-control border-0 bg-transparent text-white"
        onSubmit={[Function]}
      >
        <div
          className="mb-3"
        >
          <label
            className="form-label"
            htmlFor="emailUser"
          >
            Email :
          </label>
          <input
            autoFocus={true}
            className="form-control border-0 border-bottom bg-transparent text-white"
            id="emailUser"
            name="email"
            onChange={[Function]}
            placeholder="Ingrese e-mail"
            type="email"
            value=""
          />
        </div>
        <div
          className="mb-3"
        >
          <label
            className="form-label"
            htmlFor="passUser"
          >
            Contraseña:
          </label>
          <input
            className="form-control border-0 border-bottom bg-transparent text-white"
            id="passUser"
            name="password"
            onChange={[Function]}
            placeholder="Ingrese contraseña"
            type="password"
            value=""
          />
        </div>
        <button
          className="btn btn-dark w-100 mt-4"
          type="submit"
        >
          Login
        </button>
      </form>
      <p
        className="text-center fs-6 text-white"
      >
        ¿No tienes cuenta? Haz click en 
        <br />
        <a
          href="/register"
          onClick={[Function]}
        >
           Registrarme 
        </a>
      </p>
      <p
        className="fs-6 mt-3 text-center"
      >
        Todos los derechos reservados 
        <br />
        ©Hero App
      </p>
    </div>
  </div>
</div>
`;
